spec/models/bag_man_request_spec.rb

```diff
@@ -59,31 +59,42 @@
     end
 
     context "when not cancelled" do
+      let(:replication) {
+        mock = double(:replication_transfer)
+        allow(mock).to receive(:cancel!)
+        allow(mock).to receive(:from_node).and_return FromNode.new("fake_namespace")
+        mock
+      }
+      let(:request) {
+        request = Fabricate(:bag_man_request)
+        allow(request).to receive(:replication_transfer).and_return(replication)
+        request
+      }
       before(:each) do
-        @repl_mock = double(:replication_transfer)
-        allow(@repl_mock).to receive(:cancel!)
-        allow(@repl_mock).to receive(:from_node).and_return FromNode.new("fake_namespace")
-        @request = Fabricate(:bag_man_request)
-        allow(@request).to receive(:replication_transfer).and_return(@repl_mock)
         allow(Client::Repl::CancelJob).to receive(:perform_later)
       end
       it "sets the cancel_reason" do
-        @request.cancel!('testing')
-        expect(@request.reload.cancel_reason).to eql('testing')
+        request.cancel!('testing')
+        expect(request.reload.cancel_reason).to eql('testing')
       end
       it "calls replication_transfer.cancel!" do
-        expect(@repl_mock).to receive(:cancel!)
-        @request.cancel!('testing')
+        expect(replication).to receive(:cancel!)
+        request.cancel!('testing')
       end
       it "passes the reason to replication_transfer.cancel!" do
-        expect(@repl_mock).to receive(:cancel!).with('testing')
-        @request.cancel!('testing')
+        request.cancel!('testing')
+        expect(replication).to have_received(:cancel!).with('testing', anything)
+      end
+      it "passes last_error as the detail to replication_transfer.cancel!" do
+        request.update!(last_error: "some_error")
+        request.cancel!('testing')
+        expect(replication).to have_received(:cancel!).with(anything, request.last_error)
       end
       it "enqueues a Client::Repl::CancelJob" do
         expect(Client::Repl::CancelJob).to receive(:perform_later)
-          .with(@repl_mock, "fake_namespace", "replicate",
+          .with(replication, "fake_namespace", "replicate",
             "update_replication", "ReplicationTransferAdapter")
-        @request.cancel!('testing')
+        request.cancel!('testing')
       end
```

Comment:
The assumption is you're either cancelling the replication request because (1) the to_node cancelled it on their end and you're syncing that change, or (2) you want to cancel it for some reason.  Both of these are powered by jobs or clients, where the expectation is that a job/client will run again if it encounters an exception, up to some max number of attempts.  The code doesn't make any preparations for what happens after that.